graphql_compiler/compiler/emit_cypher.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+"""Convert lowered IR basic blocks to Cypher query strings."""
+from .blocks import Fold, QueryRoot, Recurse, Traverse
+
+
+def _emit_code_from_cypher_step(cypher_step):
+    """Return a Cypher query pattern match expression corresponding to the given CypherStep."""
+    no_linked_location = cypher_step.linked_location is None
+    step_is_query_root = isinstance(cypher_step.step_block, QueryRoot)
+
+    if no_linked_location ^ step_is_query_root:
+        raise AssertionError(u'Received an illegal CypherStep. Not having a linked location is '
+                             u'allowed if and only if the step is with a QueryRoot object. {}'
+                             .format(cypher_step))
+
+    has_where_block = cypher_step.where_block is not None
+    is_optional_step = (
+        isinstance(cypher_step.step_block, Traverse) and
+        cypher_step.step_block.optional
+    )
+    if has_where_block and is_optional_step:
+        raise AssertionError(u'Received an illegal CypherStep containing an optional step together '
+                             u'with a "where" Filter block. This is a bug in the lowering code, as '
```

Comment:
Are you 100% certain the bug is always going to be due to the lowering code? 
You could have just replaced the second sentence with "The filtering should have been moved to the global operations section".

Reply:
Filters in optional locations get moved to the global operations section in the method `move_filters_in_optional_locations_to_global_operations` in `compiler/ir_lowering_cypher/__init__.py` so seeing a filter block in an optional section means that function didn't work correctly.

The `emit_code_from_ir` function gets called immediately after the lowering function is called, so we don't have a situation where code in the middle can mix it up either.